theory/theory.md

```diff
@@ -0,0 +1,362 @@
+
+## 6. What's the difference between agile and waterfall?
+
+Agile is a collection of principles and values established in 2001 by 17 technology leaders. Scrum, Lean, Six Sigma, and Kanban are among the frameworks and product delivery methods that fall under the Agile umbrella. Agile frameworks and approaches, in general, do not govern the project lifecycle, but rather give a flexible and iterative solution that allows you to change as needed.
+
+Waterfall, on the other hand, is far more linear, focused on upfront preparation with well stated requirements before a project begins. Work cascades across several project phases, much like a waterfall, as the term implies. Each phase must be completed before moving on to the next.
+
+The primary distinctions between agile and waterfall techniques are as follows.
+
+- Customer involvement over contract negotiations
+- People and engagement take precedence over process and tools.
+- Adapting to change rather than sticking to a routine
+- Overcoming extensive documentation by developing/testing solutions
+
+The main distinction between these techniques is that, whereas Agile completes projects iteratively in a cycle of sprints, Waterfall plans and follows a sequence. Waterfall forecasts how the ultimate product will turn out, whereas Agile envisions it. Agile, in a sense, leaves nothing to chance. Despite having planned everything from the start, it nonetheless allows for and adjusts to changes and unforeseen challenges that may develop as the project cycle progresses.
+
+Before the next sprint can begin, Agile relies on client collaboration and feedback, whereas Waterfall methodologies do not value customer collaboration until the final build is ready for testing. As each Agile sprint is completed and consumer feedback is received, the team is able to make quick changes.
+
+Source: https://www.aipm.com.au/blog/agile-vs-waterfall-whats-the-difference
+https://www.nilc.co.uk/agile-v-waterfall-project-management-methodology-what-are-the-differences-and-how-it-applies-to-the-im-a-celebrity-2020-show/
+
+
+## 7. What is a reduced function used for?
+
+When a function is applied on an iterable, the reduce() function returns an aggregated value as a result. Applying the function to the first pair of items in the iterable and using the resulting value with the following item in the sequence is repeated until item n.
+
+To compute the sum of a list of numbers, here is an example.
+
+So one of the ways in python is using a basic for loop:
+
+    total = 0
+    numbers = [1, 2, 3, 4]
+    for num in numbers:
+        total += num
+    
+    print(total)  # Output: 10
+    
+Here is another approach using reduce() now:
+    
+    from functools import reduce
+    total = reduce((lambda x, y: x + y), [1, 2, 3, 4])
+    
+    print(total) # Output: 10
+
+
+Source : https://python-reference.readthedocs.io/en/latest/docs/functions/reduce.html
+https://book.pythontips.com/en/latest/map_filter.html
+

```

Comment:
Q7. 10 / 10
This is no easy concept, Keiko. Congratulations on understaning this intermediate logic in Programming.